src/core/utils/color/ColorValue.ts

```diff
@@ -0,0 +1,709 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { getBoundedValue } from '../functions/numbers';
+import { ColorByName } from './ColorByName';
+
+// cspell: ignore tbgr ttbbggrr bbggrr rrggbb aabbggrr abgr rrggbb hsla lerp torgb dhue dsaturation dvalue intpart fractpart cyanish
+
+// portions adapted from Three.js Copyright © 2010-2017 three.js authors
+
+const scratchBytes = new Uint8Array(4);
+const scratchUInt32 = new Uint32Array(scratchBytes.buffer);
+
+/**
+ * A color defined by Red, Green, and Blue
+ * r,g,b ranges are integer values 0 - 255
+ * @public
+ */
+export type RgbColor = {
+  r: number;
+  g: number;
+  b: number;
+};
+
+/**
+ * A color defined by Hue, Saturation, and Lightness.
+ */
+export type HslColor = {
+  /** 0-360 */
+  h: number;
+  /** 0-100 */
+  s: number;
+  /** 0-100 */
+  l: number;
+};
+
+/**
+ * A color defined by Hue, Saturation, and Value
+ * @public
+ */
+export type HsvColor = {
+  /** 0-360 */
+  h: number;
+  /** 0-100 */
+  s: number;
+  /** 0-100 */
+  v: number;
+};
+
+/** An unsigned 32-bit integer in 0xTTBBGGRR format
+ * @public
+ */
+export type ColorDefProps = number;
+
+/** An immutable integer representation of a color.
+ *
+ * Colors are stored as 4 components: Red, Blue, Green, and Transparency (0=fully opaque). Each is an 8-bit integer between 0-255.
+ *
+ * Much confusion results from attempting to interpret those 4 one-byte values as a 4 byte integer. There are generally two sources
+ * of confusion:
+ *  1. The ordering of the Red, Green, Blue bytes; and
+ *  2. Whether to specify transparency or opacity (sometimes referred to as "alpha").
+ *
+ * ColorValue uses `0xTTBBGGRR` (red in the low byte. 0==fully opaque in high byte) internally, but it also provides methods
+ * to convert to `0xRRGGBB` (see [[getRgb]]) and `0xAABBGGRR` (red in the low byte, 0==fully transparent in high byte. see [[getAbgr]]).
+ *
+ * The [[create]] method also accepts strings in the common HTML formats.
+ *
+ * ColorValue is immutable. To obtain a modified copy of a ColorValue, use methods like [[adjustedForContrast]], [[inverse]], or [[withTransparency]]. For example:
+ * ```ts
+ *  const semiTransparentBlue = ColorValue.blue.withTransparency(100);
+ * ```
+ * @public
+ */
+export class ColorValue {
+  private readonly _tbgr: number;
+
+  public constructor(tbgr: number) {
```

Comment:
I think would be great and simplify things a little, if we just accept any format in constructor (rhb, hsv, hsl, stringed of those, tbgr??). Would be also great, if `ColorValue` would use rgba (or hsv?? which is easier @mayank99) as base for all the calculations, as we all are more familiar with those color schemes than tbgr (mb not familiar, but jus used to that order than inversed :D). I think we could have method to get tbgr if needed (even though users could just use rgb for that and inverse).

Reply:
using `type Color = HslColor | RgbColor | HsvColor | string` from https://github.com/iTwin/iTwinUI-react/pull/346#discussion_r729926621, the constructor would change to:
```ts
public contructor(color: Color) {}
```

not sure which format would be easier for maintaining internal state, but I think rgba would be easier to avoid a full rewrite